Marlin/i2cEncoder.cpp

```diff
@@ -0,0 +1,1173 @@
+            float threshold = constrain((speed / 50),1,50) * get_error_correct_threshold();
+          #else
+            float threshold = get_error_correct_threshold();
+          #endif
+
+          //check error
+          #if ENABLED(ERROR_ROLLING_AVERAGE)
+            ((errArrayIndex >= ERROR_ARRAY_SIZE - 1) ? (errArrayIndex = 0) : errArrayIndex++);
+            errorArray[errArrayIndex] = get_axis_error_steps(false);
+            double  sum = 0;
+            double diffSum = 0;
+            for (int i=0;i<ERROR_ARRAY_SIZE;i++) {
+              sum += errorArray[i];
+              if (i > 0) diffSum += abs(errorArray[i-1]-errorArray[i]);
+            }
+            long error = (long)(sum/(ERROR_ARRAY_SIZE + 1)); //calculate average for error
+            
+          #else
+            long error = get_axis_error_steps(false);
+          #endif
+          
+          //SERIAL_ECHO("Axis err*r steps: ");
+          //SERIAL_ECHOLN(error);
+
+          #if defined(AXIS_ERROR_THRESHOLD_ABORT)
+            if(abs(error) > AXIS_ERROR_THRESHOLD_ABORT * planner.axis_steps_per_mm[encoderAxis]) {
+              //kill("Significant Error");
+              SERIAL_ECHO("Axis error greater than set threshold, aborting!");
+              SERIAL_ECHOLN(error);
+              delay(5000);
+            }
+          #endif
+
+//          switch(get_error_correct_method()) {
+//            case ECM_MICROSTEP: 
+                #if ENABLED(ERROR_ROLLING_AVERAGE) 
+                  if(errArrayIndex == 0) {
+                    // in order to correct for "error" but avoid correcting for noise and non skips
+                    // it must be > threshold and have a difference average of < 10 and be < 2000 steps
+                    if(abs(error) > threshold * planner.axis_steps_per_mm[encoderAxis] && diffSum < 10*(ERROR_ARRAY_SIZE-1) && abs(error) < 2000) { //Check for persistent error (skip)
+                      SERIAL_ECHO(axis_codes[get_axis()]);//Verbose JDK
+                      SERIAL_ECHO(" diffSum: ");
+                      SERIAL_ECHO(diffSum/(ERROR_ARRAY_SIZE-1));
+                      SERIAL_ECHO(" err detected: ");    //Verbose JDK                
+                      SERIAL_ECHO(error / planner.axis_steps_per_mm[encoderAxis]);
+                      SERIAL_ECHOLN("mm");
+                      SERIAL_ECHOLN("Correcting");
+                      thermalManager.babystepsTodo[encoderAxis] = -lround(error);
+                    }
+                  }
+                #else
+                  if(abs(error) > threshold * planner.axis_steps_per_mm[encoderAxis]) {
+                    //SERIAL_ECHOLN(error);
+                    //SERIAL_ECHOLN(position);
+                    //thermalManager.babystepsTodo[encoderAxis] -= STEPRATE * sgn(error);
+                    thermalManager.babystepsTodo[encoderAxis] = -lround(error/2);
+                  }
+                #endif
+/*              break;
+            case ECM_PLANNER:
```

Comment:
We have found that the babystepping method works very well and provides an immediate adjustment to the position. I think the other methods can be safely removed. @chrisbarr what do you think?